src/io/bgwriter.cpp

```diff
@@ -0,0 +1,74 @@
+//
+// Created by Sameer Raj on 05/08/24.
+//
+
+#include "../../headers/io/bgwriter.h"
+#include "../../headers/io/io.h"
+#include "../../headers/configs/constants.h"
+
+#include <iostream>
+#include <thread>
+#include <chrono>
+
+namespace bgwriter {
+
+    void flushDirtyPages(const BufferDescriptors& bufferDescriptors,
+                        const BufferId victimBuffer,
+                        const std::vector<Page>& bufferPool) {
+        BufferId currentBuffer = victimBuffer;
+        int flushedPages = 0;
+
+        for (int i = 0; i < BUFFER_SLOTS && flushedPages < MAX_FLUSH_PAGES; ++i) {
+            BufferDescriptor* bufferDesc = bufferDescriptors[currentBuffer];
+            if(bufferDesc == nullptr) {
+                currentBuffer = (currentBuffer + 1) % BUFFER_SLOTS;
+                continue;
+            }
+            // pin buffer, so does not get evicted
+            buffer::pin(bufferDesc);
+            if (bufferDesc->isDirty.load()) {
+                // since read lock taken, no one can write to this buffer
+                lock::getReadLock(bufferDesc->contentLock);
+
+                if (bufferDesc->isDirty.load()) {
```

Comment:
Is this if needed ? If a buffer was dirty and some other thread still writes to it, it still remains dirty. For the isDirty variable we just need any one thread dirtying the buffer to report that the buffer was touched with and that would be enough.

Reply:
it might be that the other thread has taken the write lock and then proceeds to write,
consider this scenario: 
1. write lock taken to modify the page,
2.  change the page headers like empty space -= valueSize
3. then write in the space
if the bgwriter writes between step 2 and 3, the disk will not be consistent with the real "page". Its better to restrict other thread to write.

Reply:
Also this: if (bufferDesc->isDirty.load()) {
dual checking can be removed I think, but not the lock